<h2>Multiple Linear Regression</h2>
<p>Using the Appliances Energy Data</p>

In [1]:
#importing the necessary libraries
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
#importing the dataset
mydata=pd.read_csv('energydata_complete.csv')
mydata

date  Appliances  lights         T1       RH_1  \
0      2016-01-11 17:00:00          60      30  19.890000  47.596667   
1      2016-01-11 17:10:00          60      30  19.890000  46.693333   
2      2016-01-11 17:20:00          50      30  19.890000  46.300000   
3      2016-01-11 17:30:00          50      40  19.890000  46.066667   
4      2016-01-11 17:40:00          60      40  19.890000  46.333333   
5      2016-01-11 17:50:00          50      40  19.890000  46.026667   
6      2016-01-11 18:00:00          60      50  19.890000  45.766667   
7      2016-01-11 18:10:00          60      50  19.856667  45.560000   
8      2016-01-11 18:20:00          60      40  19.790000  45.597500   
9      2016-01-11 18:30:00          70      40  19.856667  46.090000   
10     2016-01-11 18:40:00         230      70  19.926667  45.863333   
11     2016-01-11 18:50:00         580      60  20.066667  46.396667   
12     2016-01-11 19:00:00         430      50  20.133333  48.000000   
13     2016-01-11 19:10:00         250      40  20.260000  52.726667   
14     2016-01-11 19:20:00         100      10  20.426667  55.893333   
15     2016-01-11 19:30:00         100      10  20.566667  53.893333   
16     2016-01-11 19:40:00          90      10  20.730000  52.660000   
17     2016-01-11 19:50:00          70      30  20.856667  53.660000   
18     2016-01-11 20:00:00          80      30  20.890000  51.193333   
19     2016-01-11 20:10:00         140      40  20.890000  49.800000   
20     2016-01-11 20:20:00         120      20  20.890000  48.433333   
21     2016-01-11 20:30:00         190      40  20.963333  47.633333   
22     2016-01-11 20:40:00         110      40  21.033333  47.063333   
23     2016-01-11 20:50:00         110      40  21.100000  46.596667   
24     2016-01-11 21:00:00         110      30  21.133333  46.060000   
25     2016-01-11 21:10:00         110      20  21.200000  45.800000   
26     2016-01-11 21:20:00         100      30  21.290000  45.900000   
27     2016-01-11 21:30:00         100      20  21.356667  45.826667   
28     2016-01-11 21:40:00         100      20  21.390000  45.690000   
29     2016-01-11 21:50:00         100      20  21.500000  45.333333   
...                    ...         ...     ...        ...        ...   
19705  2016-05-27 13:10:00         280       0  25.033333  48.363333   
19706  2016-05-27 13:20:00         240       0  25.166667  48.156667   
19707  2016-05-27 13:30:00         250       0  25.323333  47.930000   
19708  2016-05-27 13:40:00         220       0  25.390000  47.656667   
19709  2016-05-27 13:50:00         230       0  25.500000  47.133333   
19710  2016-05-27 14:00:00         190      10  25.500000  47.060000   
19711  2016-05-27 14:10:00         160       0  25.600000  46.990000   
19712  2016-05-27 14:20:00          80       0  25.600000  46.730000   
19713  2016-05-27 14:30:00          60       0  25.566667  46.633333   
19714  2016-05-27 14:40:00          60       0  25.500000  46.360000   
19715  2016-05-27 14:50:00          60       0  25.500000  46.060000   
19716  2016-05-27 15:00:00          60       0  25.500000  45.933333   
19717  2016-05-27 15:10:00          70       0  25.500000  45.760000   
19718  2016-05-27 15:20:00          80       0  25.500000  45.626667   
19719  2016-05-27 15:30:00          80       0  25.500000  45.590000   
19720  2016-05-27 15:40:00          70       0  25.500000  45.522500   
19721  2016-05-27 15:50:00         100       0  25.500000  45.633333   
19722  2016-05-27 16:00:00         100       0  25.500000  45.730000   
19723  2016-05-27 16:10:00          90       0  25.500000  45.790000   
19724  2016-05-27 16:20:00         100       0  25.500000  45.933333   
19725  2016-05-27 16:30:00         220       0  25.426667  46.060000   
19726  2016-05-27 16:40:00         180       0  25.500000  46.530000   
19727  2016-05-27 16:50:00         120       0  25.500000  47.456667   
19728  2016-05-27 17:00:00         110       0  25.600000  47.

In [3]:
#dropping the columns 'date' and 'lights' from the dataset
my_new_data=mydata.drop(columns=['lights', 'date'])
my_new_data

Appliances         T1       RH_1         T2       RH_2         T3  \
0              60  19.890000  47.596667  19.200000  44.790000  19.790000   
1              60  19.890000  46.693333  19.200000  44.722500  19.790000   
2              50  19.890000  46.300000  19.200000  44.626667  19.790000   
3              50  19.890000  46.066667  19.200000  44.590000  19.790000   
4              60  19.890000  46.333333  19.200000  44.530000  19.790000   
5              50  19.890000  46.026667  19.200000  44.500000  19.790000   
6              60  19.890000  45.766667  19.200000  44.500000  19.790000   
7              60  19.856667  45.560000  19.200000  44.500000  19.730000   
8              60  19.790000  45.597500  19.200000  44.433333  19.730000   
9              70  19.856667  46.090000  19.230000  44.400000  19.790000   
10            230  19.926667  45.863333  19.356667  44.400000  19.790000   
11            580  20.066667  46.396667  19.426667  44.400000  19.790000   
12            430  20.133333  48.000000  19.566667  44.400000  19.890000   
13            250  20.260000  52.726667  19.730000  45.100000  19.890000   
14            100  20.426667  55.893333  19.856667  45.833333  20.033333   
15            100  20.566667  53.893333  20.033333  46.756667  20.100000   
16             90  20.730000  52.660000  20.166667  47.223333  20.200000   
17             70  20.856667  53.660000  20.200000  47.056667  20.200000   
18             80  20.890000  51.193333  20.200000  46.330000  20.200000   
19            140  20.890000  49.800000  20.200000  46.026667  20.166667   
20            120  20.890000  48.433333  20.200000  45.722500  20.166667   
21            190  20.963333  47.633333  20.260000  45.530000  20.200000   
22            110  21.033333  47.063333  20.290000  45.223333  20.260000   
23            110  21.100000  46.596667  20.356667  44.963333  20.290000   
24            110  21.133333  46.060000  20.426667  44.760000  20.290000   
25            110  21.200000  45.800000  20.500000  44.760000  20.390000   
26            100  21.290000  45.900000  20.533333  45.090000  20.390000   
27            100  21.356667  45.826667  20.666667  45.163333  20.390000   
28            100  21.390000  45.690000  20.700000  45.060000  20.390000   
29            100  21.500000  45.333333  20.700000  44.933333  20.390000   
...           ...        ...        ...        ...        ...        ...   
19705         280  25.033333  48.363333  26.528571  40.595714  28.496667   
19706         240  25.166667  48.156667  26.600000  40.940000  28.290000   
19707         250  25.323333  47.930000  26.600000  41.012857  28.200000   
19708         220  25.390000  47.656667  26.600000  41.036000  28.200000   
19709         230  25.500000  47.133333  26.600000  41.000000  28.200000   
19710         190  25.500000  47.060000  26.540000  41.000000  28.200000   
19711         160  25.600000  46.990000  26.512500  41.203750  28.200000   
19712          80  25.600000  46.730000  26.437143  41.384286  28.426667   
19713          60  25.566667  46.633333  26.370000  41.378000  28.566667   
19714          60  25.500000  46.360000  26.318571  41.264286  28.566667   
19715          60  25.500000  46.060000  26.350000  41.000000  28.426667   
19716          60  25.500000  45.933333  26.277143  41.000000  28.356667   
19717          70  25.500000  45.760000  26.200000  41.000000  28.290000   
19718          80  25.500000  45.626667  26.171429  41.000000  28.260000   
19719          80  25.500000  45.590000  26.100000  41.000000  28.200000   
19720          70  25.500000  45.522500  26.100000  41.051429  28.200000   
19721         100  25.500000  45.633333  26.080000  41.196000  28.133333   
19722         100  25.500000  45.730000  26.000000  41.428571  28.100000   
19723          90  25.500000  45.790000  26.000000  41.590000  28.100000   
19724         100  25.500000  45.933333  26.000000  41.652857  28.000000   
19725         220  25.426667  46.060000  26.000000  41.

In [8]:
#normalising the dataset
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
normalised_data = pd.DataFrame(scaler.fit_transform(my_new_data), columns=my_new_data.columns)
target = normalised_data['Appliances']
predictors=normalised_data.iloc[:,1:27]

In [6]:
#splitting the dataset into training and testing in the ratio 70:30 
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=42)


In [9]:
#checking for multicollinearity among the predictor variables
predictors.corr()

T1      RH_1        T2      RH_2        T3      RH_3  \
T1           1.000000  0.164006  0.836834 -0.002509  0.892402 -0.028550   
RH_1         0.164006  1.000000  0.269839  0.797535  0.253230  0.844677   
T2           0.836834  0.269839  1.000000 -0.165610  0.735245  0.121497   
RH_2        -0.002509  0.797535 -0.165610  1.000000  0.137319  0.678326   
T3           0.892402  0.253230  0.735245  0.137319  1.000000 -0.011234   
RH_3        -0.028550  0.844677  0.121497  0.678326 -0.011234  1.000000   
T4           0.877001  0.106180  0.762066 -0.047304  0.852778 -0.140457   
RH_4         0.097861  0.880359  0.231563  0.721435  0.122737  0.898978   
T5           0.885247  0.205797  0.720550  0.110409  0.888169 -0.050062   
RH_5        -0.014782  0.303258  0.029595  0.250271 -0.066355  0.375422   
T6           0.654769  0.316141  0.801186 -0.009670  0.686882  0.076833   
RH_6        -0.615045  0.245126 -0.580372  0.389933 -0.647672  0.514912   
T7           0.838705  0.021397  0.663660 -0.051422  0.847374 -0.250090   
RH_7         0.135182  0.801122  0.229212  0.690584  0.172624  0.832685   
T8           0.825413 -0.030053  0.578191 -0.041023  0.795283 -0.283228   
RH_8        -0.006441  0.736196  0.068534  0.679777  0.044427  0.828822   
T9           0.844777  0.115263  0.675535  0.054544  0.901324 -0.195270   
RH_9         0.071756  0.764001  0.157346  0.676467  0.134602  0.833538   
T_out        0.682846  0.340767  0.792255  0.033674  0.699417  0.118207   
Press_mm_hg -0.150574 -0.293957 -0.133028 -0.255646 -0.189974 -0.233274   
RH_out      -0.345481  0.274126 -0.505291  0.584911 -0.281718  0.356192   
Windspeed   -0.087654  0.204932  0.052495  0.069190 -0.100776  0.263188   
Visibility  -0.076210 -0.021057 -0.069721 -0.005368 -0.102310  0.017041   
Tdewpoint    0.571309  0.639106  0.582602  0.499152  0.645886  0.414387   
rv1         -0.006203 -0.000699 -0.011087  0.006275 -0.005194 -0.000477   
rv2         -0.006203 -0.000699 -0.011087  0.006275 -0.005194 -0.000477   

                   T4      RH_4        T5      RH_5    ...           T9  \
T1           0.877001  0.097861  0.885247 -0.014782    ...     0.844777   
RH_1         0.106180  0.880359  0.205797  0.303258    ...     0.115263   
T2           0.762066  0.231563  0.720550  0.029595    ...     0.675535   
RH_2        -0.047304  0.721435  0.110409  0.250271    ...     0.054544   
T3           0.852778  0.122737  0.888169 -0.066355    ...     0.901324   
RH_3        -0.140457  0.898978 -0.050062  0.375422    ...    -0.195270   
T4           1.000000 -0.048650  0.871813 -0.076489    ...     0.889439   
RH_4        -0.048650  1.000000  0.091812  0.352591    ...    -0.044518   
T5           0.871813  0.091812  1.000000  0.032786    ...     0.911055   
RH_5        -0.076489  0.352591  0.032786  1.000000    ...    -0.138509   
T6           0.652350  0.259047  0.629161 -0.078337    ...     0.667177   
RH_6        -0.703149  0.392178 -0.632384  0.263797    ...    -0.738940   
T7           0.877763 -0.131204  0.870624 -0.142823    ...     0.944776   
RH_7         0.043527  0.894301  0.148905  0.325808    ...     0.028055   
T8           0.796256 -0.167066  0.824981 -0.086991    ...     0.869338   
RH_8        -0.095192  0.847259  0.016388  0.359840    ...    -0.113014   
T9           0.889439 -0.044518  0.911055 -0.138509    ...     1.000000   
RH_9        -0.025549  0.856591  0.072308  0.272197    ...    -0.008683   
T_out        0.663478  0.293289  0.651321 -0.053126    ...     0.668220   
Press_mm_hg -0.075292 -0.250748 -0.170999 -0.119327    ...    -0.156828   
RH_out      -0.388602  0.336813 -0.273953  0.185941    ...    -0.318848   
Windspeed   -0.185747  0.300192 -0.145011  0.082044    ...    -0.177756   
Visibility  -0.104768  0.002636 -0.084164 -0.013133    ...    -0.103915   
Tdewpoint    0.519471  0.616509  0.588362  0.078175    ...     0.581483   
rv1         -0.001815 -0.001787 -0.005490 -0.011056    ...    -0.001227   
rv2         -0.001815 -0.001787 -0.005490 -

In [11]:
predictors.corr().style.background_gradient(cmap='coolwarm')


<p>The table, especially the figure above shows it, that most variables have a high correlation among themselves, (The darker the red, the higher the correlation), and this indicates multicollinearity</p>

<h4>Linear Regression</h4>
<p>Fitting the linear model on the normalised dataset</p>

In [12]:
from sklearn.linear_model import LinearRegression
linear_model = LinearRegression()
#fitting the model to the training dataset
linear_model.fit(x_train, y_train)


LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [13]:
#obtain predictions, evaluating the model on the test set
predicted_values = linear_model.predict(x_test)


In [16]:
from sklearn.metrics import  mean_squared_error
rmse = np.sqrt(mean_squared_error(y_test, predicted_values))
rmse


0.0875144494766171

In [44]:
from sklearn.metrics import r2_score
r2_score_linear = r2_score(y_test, predicted_values)
r2_score_linear

0.14890246319303535

<p>The coefficient of determination explains that 14.89% of the variation in the Appliances is predicted by the predictor variables</p>

<h4>Ridge regression</h4>
<p>In order to take care of the anomaly (i.e. the multicollinearity among the predictor variables)</p>

In [19]:
from sklearn.linear_model import Ridge
ridge_reg = Ridge(alpha=0.5)
ridge_reg.fit(x_train, y_train)


Ridge(alpha=0.5, copy_X=True, fit_intercept=True, max_iter=None,
   normalize=False, random_state=None, solver='auto', tol=0.001)

In [21]:
ridge_pred=ridge_reg.predict(x_test)


In [24]:
from sklearn.metrics import  mean_squared_error
rmse_ridge = np.sqrt(mean_squared_error(y_test, ridge_pred))
rmse_ridge


0.0875411824976933

In [45]:
from sklearn.metrics import r2_score
r2_score_ridge = r2_score(y_test, ridge_pred)
r2_score_ridge

0.14838241459278168

<h4>Lasso regression </h4>
<p>Correcting the anomaly of multicollinearity</p>

In [26]:
from sklearn.linear_model import Lasso
lasso_reg = Lasso(alpha=0.001)
lasso_reg.fit(x_train, y_train)

Lasso(alpha=0.001, copy_X=True, fit_intercept=True, max_iter=1000,
   normalize=False, positive=False, precompute=False, random_state=None,
   selection='cyclic', tol=0.0001, warm_start=False)

In [27]:
lasso_pred=lasso_reg.predict(x_test)


In [28]:
from sklearn.metrics import  mean_squared_error
rmse_lasso = np.sqrt(mean_squared_error(y_test, lasso_pred))
rmse_lasso


0.09358170467245137

In [29]:
from sklearn.metrics import r2_score
r2_score_lasso = r2_score(y_test, lasso_pred)
r2_score_lasso

0.026800880567125818

<h4>Elastic Net Regression</h4>
<p>Balancing between lasso and ridge regression</p>

In [33]:
from sklearn.linear_model import ElasticNet
from sklearn.datasets import make_regression
predictors, target = make_regression(n_features=26, random_state=42)
regr = ElasticNet(random_state=42)
regr.fit(x_train, y_train)


ElasticNet(alpha=1.0, copy_X=True, fit_intercept=True, l1_ratio=0.5,
      max_iter=1000, normalize=False, positive=False, precompute=False,
      random_state=42, selection='cyclic', tol=0.0001, warm_start=False)

In [34]:
from sklearn.metrics import  mean_squared_error
elastic_pred=regr.predict(x_test)
rmse_elastic = np.sqrt(mean_squared_error(y_test, elastic_pred))
rmse_elastic


0.09486157889204758

In [35]:
from sklearn.metrics import r2_score
r2_score_elastic = r2_score(y_test, elastic_pred)
r2_score_elastic

-1.1536689688007584e-06

In [41]:
def get_weights_df(linear_model, feat, col_name):
  #this function returns the weight of every feature
  weights = pd.Series(linear_model.coef_, feat.columns).sort_values()
  weights_df = pd.DataFrame(weights).reset_index()
  weights_df.columns = ['Features', col_name]
  weights_df[col_name].round(3)
  return weights_df

linear_model_weights = get_weights_df(linear_model, x_train, 'Linear_Model_Weight')
ridge_weights_df = get_weights_df(ridge_reg, x_train, 'Ridge_Weight')
lasso_weights_df = get_weights_df(lasso_reg, x_train, 'Lasso_weight')
elastic_weights_df =get_weights_df(regr, x_train,'Elastic_weight')

final_weights = pd.merge(linear_model_weights, ridge_weights_df, on='Features')
final_weights = pd.merge(final_weights, lasso_weights_df, on='Features')
final_weights= pd.merge(final_weights,elastic_weights_df, on='Features')

In [42]:
final_weights

Features  Linear_Model_Weight  Ridge_Weight  Lasso_weight  \
0          RH_2            -0.456698     -0.401134     -0.000000   
1         T_out            -0.321860     -0.250765      0.000000   
2            T2            -0.236178     -0.193880      0.000000   
3            T9            -0.189941     -0.188584     -0.000000   
4          RH_8            -0.157595     -0.156596     -0.000110   
5        RH_out            -0.077671     -0.050541     -0.049557   
6          RH_7            -0.044614     -0.046291     -0.000000   
7          RH_9            -0.039800     -0.041701     -0.000000   
8            T5            -0.015657     -0.020727     -0.000000   
9            T1            -0.003281     -0.021549      0.000000   
10          rv1             0.001540      0.001487     -0.000000   
11  Press_mm_hg             0.006839      0.006516     -0.000000   
12           T7             0.010319      0.010021     -0.000000   
13   Visibility             0.012307      0.012032      0.000000   
14         RH_5             0.016006      0.016176      0.000000   
15         RH_4             0.026386      0.024168      0.000000   
16           T4             0.028981      0.027006     -0.000000   
17    Windspeed             0.029183      0.030463      0.002912   
18         RH_6             0.038049      0.035023     -0.000000   
19         RH_3             0.096048      0.094976      0.000000   
20           T8             0.101995      0.100754      0.000000   
21    Tdewpoint             0.117758      0.076668      0.000000   
22           T6             0.236425      0.213316      0.000000   
23           T3             0.290627      0.287408      0.000000   
24         RH_1             0.553547      0.511932      0.017880   

    Elastic_weight  
0             -0.0  
1              0.0  
2              0.0  
3              0.0  
4             -0.0  
5             -0.0  
6             -0.0  
7             -0.0  
8              0.0  
9              0.0  
10            -0.0  
11            -0.0  
12             0.0  
13             0.0  
14             0.0  
15             0.0  
16             0.0  
17             0.0  
18            -0.0  
19             0.0  
20             0.0  
21             0.0  
22             0.0  
23             0.0  
24             0.0

In [50]:
#comparing the models
compare=pd.DataFrame(['Linear model', 'Ridge regression', 'Lasso regression', 'Elastic Net Regression'], columns=['Models'])
compare['R_squared']=[r2_score_linear,r2_score_ridge,r2_score_lasso,r2_score_elastic]
compare['RMSE'] = [rmse,rmse_ridge,rmse_lasso,rmse_elastic]
compare

Models  R_squared      RMSE
0            Linear model   0.148902  0.087514
1        Ridge regression   0.148382  0.087541
2        Lasso regression   0.026801  0.093582
3  Elastic Net Regression  -0.000001  0.094862

<p>it is reaized that <b>Ridge Regression</b>performed better amidst the other three. Even though <b>Linear regression</b> has better performance values, it is however not good to use since multicollinearity is confirmed in the dataset</p>